### PR TITLE
Feature/sbachmei/mic 4999 implement mypy

### PR DESCRIPTION
**Pull request by stevebachmeier**
### Description

- *Category*:  other
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-4999

### Changes and notes
 
This adds type hinting to get it to pass `mypy --strict`. It also
adds the checking to the github workflow.

Note that the commits are mostly contained to very strict scope; I'd
recommend reviewing commit at a time. The basic order of commits:
1. Fix one error type at a time
2. General cleanup
3. Get github actions to work
4. Implement a NestedDict type to replace instances of Any
5. more cleanup

NOTE: you can completely ignore the last commit which just renames
a bunch of the pytest variables (calling things 'd' and 's' is unreadable and
completely borks pdb usage)

### Testing

tests pass
builds finish

### CODE REVIEW

File: src/layered_config_tree/main.py [hunk 1]
```diff
@@ -497,7 +526,7 @@ def __setitem__(self, name, value):
             )
         self._set_with_metadata(name, value, layer=None, source=None)
 
-    def __getattr__(self, name):
+    def __getattr__(self, name: str) -> Any:
```
Comment (stevebachmeier) [inline src/layered_config_tree/main.py:529]: This is the one instance I'm unable to move away from Any. I think it should return `Union[ConfigNodeValue, "LayeredConfigTree"]` but this breaks a bunch of the unit tests in a way I'm unable to resolve. The issue seems to be that the tests are doing things like
```
lct = LayeredConfigTree(data, source)
...
lct.test_key.test_key2
```

But when restricting the return (ie, not just Any) then mypy yells b/c various NestedDictValue types don't have a getattr (e.g. str, float, int)

File: src/layered_config_tree/main.py [hunk 2]
```diff
@@ -229,10 +237,16 @@ class LayeredConfigTree:
...
+    _layers: list[str]
+    _children: dict[str, Union["LayeredConfigTree", "ConfigNode"]]
+    _frozen: bool
+    _name: str
+
     def __init__(
         self,
-        data: Union[Dict, str, Path, "LayeredConfigTree"] = None,
-        layers: List[str] = None,
+        data: Optional[Union[NestedDict, str, Path, "LayeredConfigTree"]] = None,
```
Comment (rmudambi) [inline src/layered_config_tree/main.py:248]: This type `Union[NestedDict, str, Path, "LayeredConfigTree"]` is reused in a lot of places. It might be useful to define this with a type alias?

File: src/layered_config_tree/__init__.py [hunk 3]
```diff
@@ -7,6 +7,15 @@
     __title__,
     __uri__,
 )
+
+__all__ = [
```
Comment (rmudambi) [inline src/layered_config_tree/__init__.py:11]: What does this do?

File: setup.py [hunk 4]
```diff
@@ -33,14 +33,19 @@
...
         exec(f.read(), about)
 
     with (base_dir / "README.rst").open() as f:
         long_description = f.read()
 
-    install_requirements = ["pyyaml>=5.1"]
+    install_requirements = [
+        "pyyaml>=5.1",
+        # typing extensions
+        "types-PyYAML",
```
Comment (rmudambi) [inline setup.py:46]: Should these be in a separate `dev` install block? You wouldn't need them if you were going to install this library in an environment and then use it in a notebook, for example, but I'm not sure how it would work if vivarium which uses layered_config_tree had mypy run on it.

File: tests/test_basic_functionality.py [hunk 5]
```diff
@@ -34,7 +34,7 @@ def empty_node(layers: list[str]) -> ConfigNode:
 def full_node(layers_and_values: dict[str, str]) -> ConfigNode:
     n = ConfigNode(list(layers_and_values.keys()), name="test_node")
     for layer, value in layers_and_values.items():
-        n.update(value, layer, source=None)
+        n.update(value, layer, source="test_source")
```
Comment (rmudambi) [inline tests/test_basic_functionality.py:37]: Why are you providing a source here? Do we have any tests with a None source, since the typing of sources is `Optional[str]`?

File: tests/test_basic_functionality.py [hunk 6]
```diff
@@ -251,9 +254,7 @@ def test_tree_coerce_dict() -> None:
 
 
 def test_tree_coerce_str() -> None:
-    d = """"""
     s = "test"
-    assert LayeredConfigTree._coerce(d, s) == (None, s)
```
Comment (rmudambi) [inline tests/test_basic_functionality.py:253]: Why are these lines removed?

File: src/layered_config_tree/main.py [hunk 7]
```diff
@@ -67,10 +71,10 @@ class ConfigNode:
 
     """
 
-    def __init__(self, layers: List[str], name: str):
+    def __init__(self, layers: list[str], name: str):
         self._name = name
         self._layers = layers
-        self._values = {}
+        self._values: dict[str, tuple[Optional[str], ConfigNodeValue]] = {}
```
Comment (rmudambi) [inline src/layered_config_tree/main.py:77]: When would the first value of the tuple be None?

File: src/layered_config_tree/main.py [hunk 8]
```diff
@@ -322,10 +327,12 @@ def to_dict(self) -> dict[str, Any]:
             if isinstance(child, ConfigNode):
                 result[name] = child.get_value(layer=None)
             else:
-                result[name] = child.to_dict()
+                result[name] = child.to_dict()  # type: ignore[assignment]
```
Comment (rmudambi) [inline src/layered_config_tree/main.py:333]: What's happening here?

File: src/layered_config_tree/types.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
+from typing import Mapping, Union
+
+# NOTE: py 3.9 does not support typing.TypeAlias
+# Define a nested dictionary of unknown depth for type hinting
+NestedDict = Mapping[str, "NestedDictValue"]
+# Define the leaf (ConfigNode) values (which cannot be another dictionary)
+ConfigNodeValue = Union[str, int, float, list[Union[str, int, float]]]
```
Comment (rmudambi) [inline src/layered_config_tree/types.py:7]: Maybe call this `NodeValue` to match `NestedDict` and `NestedDictValue`?

File: tests/test_basic_functionality.py [hunk 10]
```diff
@@ -237,251 +242,258 @@ def test_tree_creation(empty_tree):
...
-    d, s = {}, "test"
-    assert LayeredConfigTree._coerce(d, s) == (d, s)
-    d, s = {"key": "val"}, "test"
-    assert LayeredConfigTree._coerce(d, s) == (d, s)
-    d = {"key1": {"sub_key1": ["val", "val", "val"], "sub_key2": "val"}, "key2": "val"}
-    s = "test"
-    assert LayeredConfigTree._coerce(d, s) == (d, s)
+def test_tree_coerce_dict() -> None:
+    data: NestedDict
+    data, src = {}, "test"
```
Comment (rmudambi) [inline tests/test_basic_functionality.py:247]: I think you should just assign to these variables separately:
```suggestion
    data: NestedDict = {}
    src = "test"
```

File: src/layered_config_tree/__init__.py (hunk 3 above)
Comment (stevebachmeier) [reply src/layered_config_tree/__init__.py:11]: It solves error: `error: Module "layered_config_tree" does not explicitly export attribute "ConfigurationKeyError"  [attr-defined]`

Comment (stevebachmeier) [reply src/layered_config_tree/__init__.py:11]: My understanding is that this __all__ list limits what actually gets installed if one were to `pip isntall *`. But more to the point it is one of the ways to pass mypy. I can't say I totally understand it.

File: src/layered_config_tree/main.py (hunk 7 above)
Comment (stevebachmeier) [reply src/layered_config_tree/main.py:77]: I don't think I was ever able to track down why this might actually be None (it's the data `source`). I do know that there are a bunch of tests that pass in None. I tried changing all of those to string values but things got complicated so I gave up. I can try again if you agree the source should never be None (seems like it shouldn't eveer be)

File: src/layered_config_tree/main.py (hunk 8 above)
Comment (stevebachmeier) [reply src/layered_config_tree/main.py:338]: I gave up trying to fix that. It was throwing

`error: Incompatible types in assignment (expression has type "Mapping[str, NestedDictValue]", target has type "str | int | float | list[str | int | float]")  [assignment]
Found 1 error in 1 file (checked 13 source files)`

File: src/layered_config_tree/main.py (hunk 1 above)
Comment (stevebachmeier) [reply src/layered_config_tree/main.py:534]: I do change setattr in a later commit. I guess this is where my commit-by-commit suggestions gbreaks down since some of this stuff is general cleanup

File: tests/test_basic_functionality.py (hunk 6 above)
Comment (stevebachmeier) [reply tests/test_basic_functionality.py:253]: I think that we can't have None for data. Perhaps I'm wrong, though, in which case the typing is messed up here

File: src/layered_config_tree/main.py [hunk 11]
```diff
@@ -231,8 +232,8 @@ class LayeredConfigTree:
 
     def __init__(
         self,
-        data: Union[Dict, str, Path, "LayeredConfigTree"] = None,
-        layers: List[str] = None,
+        data: Optional[Union[Dict[str, Any], str, Path, "LayeredConfigTree"]] = None,
+        layers: Optional[List[str]] = None,
```
[the remark this answers is not in the document]
Comment (stevebachmeier) [reply src/layered_config_tree/main.py]: To be honest, I'm pretty confused by layers, source, and data and whther they are allowed to be empty or None. It might be worth just talking through a few of these comments

File: src/layered_config_tree/main.py (hunk 1 above)
Comment (stevebachmeier) [reply src/layered_config_tree/main.py:528]: Discussed that the issue is that mpyp doesn't know the returns of a chained attr. Keeping as Any for now

File: src/layered_config_tree/main.py (hunk 7 above)
Comment (stevebachmeier) [reply src/layered_config_tree/main.py]: Discussed and looks correct. Weird that source is first in the tuple, though

File: tests/test_basic_functionality.py (hunk 6 above)
Comment (rmudambi) [reply tests/test_basic_functionality.py:253]: Don't forget to re-remove these lines.

**Review by rmudambi — APPROVED** [review]
Approved assuming you fix the build.

**Review by albrja — APPROVED** [review]
